#summary AltAnalyze Included Annotations

== Annotations Provided in AltAnalyze ===

Annotations are provided for the various analyses in AltAnalyze at the level of genes, alternative exons, splicing events, microRNA binding sites and proteins. A full description of these annotations can be found in the *[http://www.altanalyze.org/help.htm full online documentation]*.

=== Gene Expression Results ===

AltAnalyze extracts useful gene annotations for determining whether a gene: 

 * Protein-coding potential
 * “Drug-able” targets (e.g. GPCR, kinase)
 * Transcriptional and splicing regulator
 * Putative microRNA targets
 * Associated WikiPathways
 * Associated Gene Ontology terms
 * UniProt annotated cellular compartment
 * Putative housekeeping genes

=== Alternative Isoform Annotations ===

Alternative isoform data is obtained through extraction of transcript data from the Ensembl and UCSC databases. This database is primarily built from Ensembl and augmented with unique exons, junctions and transcripts found in the UCSC mRNA table for each species in Ensembl. Novel isoform annotations are included for Affymetrix junction arrays (HJAY, hGlue) and imported RNASeq data. These annotations include:

 # mRNA transcripts
 # associated proteins
 # exons
 # junctions
 # exon/junction aligning protein domains (Ensembl)
 # exon/junction aligning protein motifs (UniProt)
 # exon aligning microRNA binding sites ([http://www.altanalyze.org/help.htm#microrna various])
 # [http://www.altanalyze.org/help.htm#as_predict alternative splicing events, promoters and polyadenylation] 

=== Pathway Over-representation ===

See [http://genmapp.org/go_elite/help.htm#resources GO-Elite webpage].