#summary Frequently Asked Questions
#labels Featured

=AltAnalyze Frequently Asked Questions=

===Main Analysis Options===

    * [CompatibleOS What operating systems is AltAnalyze compatible with?]
    * [CommandLineMode Can I run AltAnalyze on a remote machine without using the graphical user interface?]
    * [SpliceVisualization How can I visualize splicing and domain level changes from AltAnalyze?]
    * [CompatibleArrays What microarrays can AltAnalyze work with?]
    * [SplicingForAnyPlatform Can AltAnalyze work with exon/junction data for an unsupported platform (e.g., custom array)?]
    * [ScoringMethod What is the difference between the splicing-index, FIRMA, MiDAS, Linear Regression and ASPIRE methods?]
    * [AnnotateExternal How can I annotate my splicing results from another program in AltAnalyze?]

===Installing AltAnalyze===

    * [DBaseInstallationError Why Does AltAnalyze crash when trying to download the database when first run?]

===Data Analysis and Import===
    * [RequiredFiles What files do I need when analyzing Affymetrix arrays?]
    * [ManualGroupsCompsCreation When I have dozens of samples in my study, how can I more easily assign the samples to groups than one by one?]
    * [RNASeqExpressionCutoffs What is the recommended cutoff for expressed genes with RNA-Seq?]
    * [FilteringData Why does AltAnalyze need to filter my data?]
    * [Constitutive_vs_all Should I calculate gene expression using constitutive or all mRNA aligning probests?]
    * [MultipleComparisons How does a pair-wise alternative exon analysis differ from multiple group comparisons?]
    * [GroupsAndComps Why do I need to define groups and comparisons?]
    * [GOEliteDefault Why doesn't AltAnalyze run GO-Elite by default?]
    * [GeneArrayCSVForExon Why does AltAnalyze download the Gene 1.0 transcript annotation file for Exon 1.0 array analyses?]
    * [CrossHybProbesets How can I remove probesets with sequence that aligns to multiple genomic loci from my analysis?]
    * [TranscriptClusterCenteredResults Can AltAnalyze return transcript cluster results instead of Ensembl for the Gene 1.0 array?]
===Probeset Alignments and DomainGraph===

    * [ExonMappings How do AltAnalyze probeset-to-gene and probeset-to-exon mappings differ from those in the Affymetrix annotation files?]
    * [DGIntronMapping In AltAnalyze a probeset associated with an intron is predicted to alter protein domain composition of the protein, but in DomainGraph why don't I see the probeset?]
    * [StartingDG How do I start DomainGraph and what can I do in DomainGraph?]

===Interpreting Alternative Exon Results===
    * [RNASeqFoldsAndExpression Why don't the folds and expression values from my RNA-seq experiment don't seem to match correctly?]
    * [GeneExpCalculation What is a constitutive exon/junction?]
    * [ResultsVerification Which results should I validate?]
    * [OppositeMappings If two probesets align to one exon and have opposite splicing changes, what does this mean?]
    * [ProteinDirectionIndicator What does the (+) and (-) mean proceeding a protein domain, miRNA binding site or protein identifier?]
    * [ExonRegionIDs What does the identifier E2-3 indicate and where does this annotation come from?]
    * [ASAnnotations Where do the alternative splicing annotations provided in the column "exon annotations" come from?]
    * [DistalExonRegion What does the column "distal exon-region-ID" indicate?]
    * [AdjFoldChange What is the adjusted fold versus the non-adjusted fold change?]
    * [EventCall Why do all alternative exons have the indicator "exon-inclusion" in the column "event_call"?]
    * [GraphingExonData How can I graph my splicing scores for all exons in a gene?]
    * [ExonVsGene What is the difference between exon-level and gene-level alternative exon results?]
    * [DomainGraphNonExon When viewing AltAnalyze results in DomainGraph, why don't the identifiers match to my array/RNA-seq platform?]

===Other===
    * [AboutUs Who are you and what's your deal?]