#summary FAQ-Gene Array Transcript Annotations for Exon Analyses

===_Why does AltAnalyze download the Gene 1.0 transcript annotation file for Exon 1.0 array analyses?_===


_*Answer*_: When analyzing exon array CEL files in AltAnalyze, AltAnalyze will automatically download the appropriate array library files. Unlike other array types however, AltAnalyze downloads a different array type annotation file, the Gene 1.0 array annotation CSV file (e.g., HuGene-1_0-st-v1.na28.hg18.transcript.csv) with these library files. The reason a different array type annotation file is downloaded, is that AltAnalyze only gathers Ensembl-Gene Ontology and Ensembl-pathway annotations from this file for alternative exon analyses. These only provide annotations for the the gene-level results. This file is also not required for analyses. 

This can be confusing for users, since users often question why Gene 1.0 array annotations would be used for exon array analyses. Unlike other array types, AltAnalyze has almost all probeset annotations stored locally in its installed database and thus doesn't depend on the CSV annotations. 