#summary Instructions for Obtaining RNASeq Input Files

= Instructions for Obtaining RNASeq Input Files =

*_Please Note:_* We are actively working on ways to better integrate existing RNA-seq alignment strategies into AltAnalyze. Recommendations to our [ContactUs support desk] are always welcome.

===Downloading Sample FASTQ Data===

You are welcome to try with your own FASTQ data, but if none is available, which will be easiest for most users than downloading large datasets. For optimal junction analysis results, we recommend >50nt reads, however, we have achieved acceptable results from 35nt reads. If you currently do have your own data, example files and download instructions can be found [DownloadFASTQ here].

===Using TopHat and Bowtie===

Example Instructions for obtaining exon and junction expression files (BED format). Under Construction.

{{{
bamToBed -i accepted_hits.bam -split| coverageBed -a stdin -b /home/user/BAMtoBED/hESC_differentiation_exons.bed > /home/user/RNASeqStudy/Sample1/day0_s1__exons.bed
}}}

===Using BioScope===

Example Instructions for obtaining exon and junction expression files (TAB format). Under Construction.