#summary Instructions for Obtaining RNASeq Input Files

= Instructions for Obtaining RNASeq Input Files =

*_Please Note:_* We are actively working on ways to better integrate existing RNA-seq alignment strategies into AltAnalyze. Recommendations to our [ContactUs support desk] are always welcome.

===Downloading Sample FASTQ Data===

You are welcome to try with your own FASTQ data, but if none is available, which will be easiest for most users than downloading large datasets. For optimal junction analysis results, we recommend >50nt reads, however, we have achieved acceptable results from 35nt reads. If you currently do have your own data, you can download one of the below recommended studies from [http://www.ncbi.nlm.nih.gov/sra SRA]:
 
 * ERP000570 GSE21860 GSE27218 (Mus musculus) 
 * GSE24447 (Homo sapiens)

These studies contain multiple replicates for multiple in vivo/in vitro conditions. To obtain FASTQ sequence:
 # Search for the above accession numbers at SRA.
 # For each sample, select the experiment accession number on the right side of the page (under Total:).
 # Select the button "Download" half way down the page.
 # On the next page select "FASTQ", de-select "filtered" and select "save".
 # Repeat for each sample (often 1-7GB files) and rename after download.

These files will need to be moved to a suitable machine for (ideally multiple processors with 8GB plus RAM) for sequence alignment. For paired-end data, both paired ends are stored sequentially in a file, for example:
@SRR037165.2.1 FC30B11:1:1:256:1595.1 length=35
@SRR037165.2.2 FC30B11:1:1:256:1595.2 length=35


===Using TopHat and Bowtie===

Example Instructions for obtaining exon and junction expression files (BED format). Under Construction.

{{{
bamToBed -i accepted_hits.bam -split| coverageBed -a stdin -b /home/user/BAMtoBED/hESC_differentiation_exons.bed > /home/user/RNASeqStudy/Sample1/day0_s1__exons.bed
}}}

===Using BioScope===

Example Instructions for obtaining exon and junction expression files (TAB format). Under Construction.