#summary FAQ-Analyze Transcript Clusters Instead of Ensembl Genes

==_Can AltAnalyze return transcript cluster results instead of Ensembl for the Gene 1.0 array?_==

_*Answer*_: In addition to obtaining Ensembl gene centric gene expression results, AltAnalyze can be parameterized to return transcript cluster centric expression results. This can be down by following the steps in the below sections

===_Background_===
The Affymetrix Gene 1.0 array contains multiple probesets per transcript cluster, with typically one probeset for each well annotated exon region. These array designs allows one to obtain transcript cluster expression values for fairly well annotated genes that consider all associated probesets or alternative exon analyses that consider each individual probeset relative to a conservative gene expression estimate (constitutive exon or all exon summary). By default, AltAnalyze performs the latter analysis, providing probeset-level alternative exon results, nearly identical to the Affymetrix Exon 1.0 array and provides and gene expression (transcriptional activity) summary values based not on the transcript cluster ID, but rather the mean of probesets that align to an Ensembl gene (constituent exons). 

===_Analysis: Starting from CEL files_===

 # Move all of the CEL files from your study of interest to a single folder.
 # Go to: http://www.affymetrix.com/support/technical/byproduct.affx? and select “Mouse Genome Arrays”, then “Mouse Gene 1.0 ST Array” and then "Unsupported Mouse Gene 1.0 ST Array CDF Files”. This will download the CDF archive to your computer. You must then unzip this file onto your hard-drive.
 # Start AltAnalyze
 # Select Affymetrix from platform, your species and array type as Affymetrix expression array (not Gene Array)
 # Select “Process CEL files”
 # Enter a name for the experiment, select the folder containing these CEL files and the same folder as the output directory.
 # Next, you will be prompted to “Use Original Selected” or “Use AltAnalyze Recommended” for your array type. Selecting “Use Original Selected” will allow you to use your own CDF to obtain transcript cluster probeset expression as opposed to exon-level expression.
 # Select the folder you extracted in step 2 and the CDF file within it. There is no need to select an annotation file as AltAnalyze now has all the annotations you need built in.
 # Select proceed on the menu with additional parameters to accept the defaults
 # When prompted to enter group name for each sample.
 # Assign group comparisons as your experimental on the left and control on the right from the pull-down.
 # Once finished open the file with the prefix “DATASET” in the folder ExpressionOutput in the output directory you designated.
 # This is your initial results re-annotated with AltAnalyze gene annotations.

===_Analysis: Starting Already Summarized Transcript Cluster Data_===

 # Save your Excel spreadsheet as a tab-delimited text (.txt) file with the following fields:
   * Transcript cluster probeset ID
   * Expression values
   * Fold changes (log and non-long)
   * P-values
 # Start AltAnalyze
 # Select Affymetrix from platform, your species and array type as Affymetrix expression array (not Gene Array)
 # Select “Process Expression files” rather than “Process CEL files”
 # Select this text file as input and the same directory as the output location
 # Select proceed on the menu with additional parameters to accept the defaults
 # When prompted to enter group name for each sample, just enter a or b at random into for the different samples. You will ignore these resulting groups and statistics later.
 # Assign group a versus b and continue
 # Once finished open the file with the prefix “DATASET” in the folder ExpressionOutput in the output directory you designated.
 # This is your initial results re-annotated with AltAnalyze gene annotations. You can delete the columns added by AltAnalyze by this point.