#summary RNASeq data description

==Introduction==

RNA sequencing data or RNA-seq, produced from whole transcriptome shot-gun sequencing, provides digital reads for short sequences from the transcriptome of cells or tissues. Several technologies are available for the amplification and detection of RNA-seq reads, including single read, paired-end and pulse-read sequences. 

==Analysis in AltAnalyze==

AltAnalyze version 2.0 provides new methods for analyzing RNA-seq data, post genomic/transcriptome alignment. Although the initial processing of RNA-seq data is distinct from microarray data in AltAnalyze, both RNA-seq data is analyzed using the same downstream algorithms of junction sensitive splicing arrays (e.g., ASPIRE and Linear Regression reciprocal junction analyses). Databases for dozens of species are currently being built and maintained by the AltAnalyze staff, however, users can collaborate with the AltAnalyze development team or [BuildingDatabases customize AltAnalyze source code] to add compatibility for new species. 

To analyze your RNA-seq data, you must first align your read sequences for each experimental sample to junctions using external software. These packages include [http://derisilab14.ucsf.edu/software/hmmsplicer/ HMMSplicer] and [http://tophat.cbcb.umd.edu/manual.html TopHat]. These packages produced junction alignments and corresponding junction counts in the form of a BED format file (e.g., junction.bed). These files are used as input to AltAnalyze, similar to Affymetrix CEL files. For more information, see the [Tutorial_AltExpression_RNASeq AltAnalyze Analysis Tutorial].