#summary Tutorial 3 - Analyzing RNASeq Splicing Data

= Tutorial 3 - Analyzing RNASeq Alternative Junctions =

==Introduction==
The primary use of AltAnalyze is to evaluate alternative splicing, alternative promoters or other forms of alternative gene regulation. To do this, AltAnalyze imports pre-aligned exon-junctions coordinates and read counts,  calculates a reciprocal junction alternative exon-score (ASPIRE and LinearRegression), assigns exon/intron/splicing annotations to these results and further assesses protein, protein domain and microRNA binding site changes for associated isoforms.  AltAnalyze makes this process relatively easy, with the user only required to download and extract the program and provide one set of basic files. In the following tutorial we will walk through these steps using a sample dataset. Note: If analyzing [JAY junction arrays] as opposed to RNASeq data, the analysis options and result files are nearly identical.

==Downloading Sample Data==

Proceed [RNASeq_sample_data here].

==Installing AltAnalyze and Saving Your Data==
AltAnalyze version 2.0 can be downloaded for multiple operating systems from [http://www.altanalyze.org http://www.altanalyze.org]. Once you have downloaded the compressed archive to your computer, extract it to an accessible folder on hard-drive (e.g., your user account). In addition to AltAnalyze, Cytoscape and DomainGraph are automatically downloaded when a species database is first installed. See the 2nd to last step in the Running AltAnalyze instructions for how to immediately start DomainGraph after generating results.

==Creating a Comparison and Groups File (OPTIONAL)==
If your dataset has over 30 BED files or dozens of groups, it may save you time to make the groups and comps files in advance. Although not recommended when working with this sample dataset, go [ManualGroupsCompsCreation here] if this applies to your own dataset.

==Running AltAnalyze==
Now you are ready to process your input files and obtain alternative exons with alternative splicing and functional annotations. To proceed:

    * Open the AltAnalyze_v2release program folder and select the binary file “AltAnalyze”. In Windows, this file has the extension “.exe”. If you are having problems starting AltAnalyze, you can also start the program directly from the [RunningSourceCode source code].
    * ([http://altanalyze.org/image2/introduction.jpg AltAnalyze: Introduction]) In the resulting introduction window, select “Begin Analysis”.
    * ([http://altanalyze.org/image2/DB_download.jpg AltAnalyze: Download Databases]) When starting AltAnalyze for the first time you will be prompted to download databases. In this initial prompt, select "Continue", unless you are working on a computer without an internet connection. If the latter applies, [AltAnalyzeOffline click here]. Otherwise, select the species "Homo sapiens" from the species select drop down menu (leaving the default Ensembl version) and then "Continue".
    * ([http://altanalyze.org/image2/Select_RNASeq_pulldown.jpg AltAnalyze: Select Species and Platform Type]) Select the species “Homo sapiens”, the platform "RNASeq" and “Continue”.
    * ([http://altanalyze.org/image2/Select_Analysis_Method_RNASeq.jpg AltAnalyze: Select Analysis Method]) Select the “Process Junction files” button and "Continue".  
    * ([http://altanalyze.org/image2/Select_BED_files.jpg AltAnalyze: Select BED files for Import]) For dataset name, type in "hESC_differentiation", select the directory containing the BED files (make sure these have been extracted from the TAR file and Gzip files) and then select an empty directory to save the results to.
    * ([http://altanalyze.org/image2/ExpressionAnalysis_RNASeq.jpg AltAnalyze: Expression Analysis Parameters]) Accept the default parameters by clicking “Continue”.  These options can be modified later on, if you wish to increase or decrease the stringency of the analysis.
    * ([http://altanalyze.org/image2/AltExonAnalysis_RNASeq.jpg AltAnalyze: Alternative Exon Analysis Parameters]) Accept the default parameters by clicking “Continue”. 
    * ([http://altanalyze.org/image2/groups.jpg AltAnalyze: Assign BED files to a Group]) Type in the group name for each sample (first three are hESC and second three are neural_precursors).
    * ([http://altanalyze.org/image2/pairwise_comps.jpg AltAnalyze: Establish All Pairwise Comparisons]) Selects the groups to compare (neural_precursors vs. hESC).
    * A new window will appear that displays the progress of the analysis. Analysis of the sample dataset should take approximately 5 minutes, but can take longer depending upon the machine and operating system. When finished, AltAnalyze will display a new pop-up window, informing the user that the analysis is complete. If analyzing multiple experimental comparisons, these analyses will be run in succession.
    * ([http://altanalyze.org/image/R1.jpg AltAnalyze: Summary Results]) For the last comparison run, overview results are printed to the user interface, including the number of genes with predicted alternative exons, number of genes with predicted domains/motifs and microRNA binding sites affected and number of genes examined. This window will also have the option to [http://altanalyze.org/image/R2.jpg Start DomainGraph in Cytoscape]. Select this option to immediately open Cytoscape and load your results in DomainGraph (once the DomainGraph database is installed). For more information,  [DomainGraph click here].
    * ([RunningGOElite Performing GO-Elite Analysis within AltAnalyze]) To determine if there are biological pathways (WikiPathways) or Gene Ontology (GO) categories that contain a disproportionate number of alternatively regulated genes, you can use the GO-Elite option in AltAnalyze. [GOElite GO-Elite] performs typical over-representation analysis (ORA), filter-out redundant GO-terms and pathways for publication ready tables, summarize all AltAnalyze statistics at the pathway level and easily view gene annotations for each pathways.

==Interpreting the Results==
When AltAnalyze was running it produced a number of output files, most to the folder AltResults/AlternativeOutput in the user output directory. These include:

  # ExpressionOutput/DATASET-YourExperiment.txt
  # AltResults/AlternativeOutput/YourComparison-ASPIRE-exon-inclusion-results.txt
  # AltResults/AlternativeOutput/YourComparison-ASPIRE-exon-inclusion-GENE-results.txt
  # AltResults/AlternativeOutput/YourComparison-ASPIRE-ft-domain-zscores.txt
  # AltResults/AlternativeOutput/YourComparison-ASPIRE-miRNA-zscores

These files are tab-delimited text files that can be opened in a spreadsheet program like Microsoft Excel, OpenOffice or Google Documents. In addition to these files, similar files will be produced with the algorithm "splicing-index" (replaces the filename ASPIRE above with splicing-index). These are similar format files with single junction results (as opposed to reciprocal-junction pairs). These results allow users to examine the independent regulation of exon junctions.

*File #1* reports gene expression values for each sample and group based on junctions present in your input BED files. The values are derived from junctions that align to regions of a gene that are common to all transcripts and thus are informative for transcription (unless the option "known junctions" is selected – see “Select expression analysis parameters”, above) and expressed above specified background levels (minimum group average read count).  Along with the raw gene expression values (mean read counts), statistics for each indicated comparison (mean expression, folds, t-test p-values) will be included along with gene annotations from Ensembl, including putative microRNA binding sites.  This file is analogous to the results file you would have with a typical microarray experiment and is saved to the folder “ExpressionOutput”.

Results from *files #2-5* are produced from all junctions that may suggest alternative splicing, alternative promoter regulation, or any other variation indicated by a reciprocal junction analysis for that gene. Each set of results correspond to a single pair-wise comparison (e.g., cancer vs. normal) and will be named with the group names you assigned (groups file). If analyzing a multiple groups, the two groups with the largest difference in reciprocal junction scores will be reported along with the conditions these occur in.

*File #2* reports reciprocal junctions that are alternatively expressed, based on the user defined ASPIRE or LinearRegression scores  and p-values. For each reciprocal junction has several statistics, gene annotations and functional predictions provided. A detailed description of all of the columns in this file is provided [http://altanalyze.org/image/AltAnalyze_column_annotations.xls here].

*File #3* is a summarization of reciprocal-junction results at the gene level from file #2. In addition to this summary, Gene Ontology terms and WikiPathways for that gene are reported.

*Files #4 and #5* report over-representation results for protein domains (or other protein features) and microRNA-binding sites, predicted to be regulated by AltAnalyze.  These files include over-representation statistics and genes associated with the different domains or features¸ predicted to be regulated.

More information about these files can be found in the [http://www.altanalyze.org/help_main.htm AltAnalyze ReadMe] (section 2.3).

==Full Directory Tree of Output Files==

After you run AltAnalyze, the following directory tree and set of files will be generated in the folder that you specified for output. For this example, the species is "species," and we assume there were two comparisons made, between groups A and B, and between groups B and C.

   * AltAnalyze_report.log
   * AltExpression/
      * FullDatasets/
         * RNASeq/
            * species/
               * YourExperiment.txt
      * RNASeq/
         * species/
            * A_vs_B.txt
            * B_vs_C.txt
      * pre-filtered/
         * expression/
            * A_vs_B.txt
            * B_vs_C.txt
   * AltResults/
      * AlternativeOutput/
         * ...
      * DomainGraph/
         * ...
      * RawSpliceData/
         * species/
            * ASPIRE/
               * A_vs_B.txt
               * B_vs_C.txt
   * ExpressionInput/
      * comps.YourExperiment.txt
      * exp.YourExperiment-steady-state.txt
      * exp.YourExperiment.txt
      * groups.YourExperiment.txt
      * stats.YourExperiment.txt
   * ExpressionOutput/
      * DATASET-YourExperiment.txt
      * GenMAPP-YourExperiment.txt
   * GO-Elite/
      * denominator/
         * ...
      * GO-Elite_report.log
      * GO-Elite_results/
         * CompleteResults/
            * ORA/
               * ...
            * ORA_pruned/
               * ...
         * pruned-gene-associations.txt
         * pruned-results_z-score_elite.txt
      * input/
         * ...

==Visualizing AltAnalyze Results in DomainGraph==
The text file results produced by AltAnalyze can be directly used as input in the protein domain and microRNA binding site visualization program, DomainGraph. DomainGraph is a plugin for the Java program Cytoscape which can be immediately opened from AltAnalyze. Rather than visualizing junctions, however, DomainGraph currently only supports exon visualization. RNASeq highlighted exons (identified by reciprocal exon-junctions) are associated with Affymetrix Exon 1.0 identifiers to be loaded in DomainGraph.

==Visualizing Over-Represented GO terms and Pathways==
Once over-represented pathways have been found or before doing this analysis, you can see which genes on which pathways are alternatively regulated in the program PathVisio or GenMAPP 2.1. PathVisio is a cross-platform analysis program, while GenMAPP is restricted to Windows. Both tools are easy use and have access to a large archive of curated pathways. An input file for either PathVisio or GenMAPP is found in the directory "ExpressionOutput" with the prefix "GenMAPP-". For making pathways, PathVisio or WikiPathways is recommended, since these resources produce superior pathway content (valid interactions between genes and metabolite IDs) in the same format (gpml). PathVisio can also export pathways to the GenMAPP format. A PathVisio tutorial can be [http://www.pathvisio.org/wiki/PathVisioTutorials found here], while a GenMAPP tutorial can be [http://www.genmapp.org/tutorial_v2.html found here].