#summary Tutorial 4 - De Novo Identification of Sample Groups

= Tutorial 4 - De Novo Identification of Sample Groups =

A challenging bioinformatics challenge is the unbiased identification of both major and sub-populations of bulk or single-cell transcriptomes where no obvious reference gene or splice signatures exist. In addition, while transcriptome samples are often annotated as belonging to one or more groups, miss-classified samples or other unexpected covariants may also result in atypical variation that is difficult to characterize. A new set of algorithms have been implemented in AltAnalyze to identify the most coherent gene and sample expression signatures, prior to group comparison analyses. These new methods become particularly important when considering unknown single-cell RNA-Seq profiles and tumor subtype classification. In addition to gene expression, the software allows for the highly accurate and unbiased identification of highly similar alternative splicing signatures.

== Algorithms ==

== Input Data Formats ==

These analyses are compatible with all regular supported file formats in AltAnalyze (e.g., RNA-Seq bed, Affymetrix CEL, Agilent feature extraction, non-zero FPKM txt files). For RNA-Seq we recommend using either junction.bed files directly (TopHat, SpliceMap), TCGA junction_expression.txt files and/or BedTools produced exon.bed files. For more information on these inputs, go http://code.google.com/p/altanalyze/wiki/Tutorial_AltExpression_RNASeq here].
